MNIST/MNIST.swift

```diff
@@ -12,115 +12,86 @@
-            let gradients = MNISTParameters(w1: dw1, w2: dw2, b1: db1, b2: db2)
-
-            // Update parameters.
-            parameters.update(withGradients: gradients) { param, grad in
-                param -= grad * learningRate
-            }
-
-            // Calculate the sigmoid-based cross-entropy loss.
-            // TODO: Use softmax-based cross-entropy loss instead. Sigmoid
-            // cross-entropy loss treats class labels as independent, which is
-            // unnecessary for single-label classification tasks like MNIST.
-            // Sigmoid cross-entropy formula from:
-            // https://www.tensorflow.org/api_docs/python/tf/nn/sigmoid_cross_entropy_with_logits
+let epochCount = 20
+let batchSize = 20
 
-            // let part1 = max(predictions, 0) - predictions * labels
-            // let part2 = log(1 + exp(-abs(predictions)))
-            // loss = ((part1 + part2).sum(squeezingAxes: 0, 1) / batchSize).scalarized()
+func minibatch<Scalar>(in x: Tensor<Scalar>, at index: Int) -> Tensor<Scalar> {
+    let start = Int32(index * batchSize)
+    return x[start..<start+Int32(batchSize)]
+}
 
-            // Update number of correct/total guesses.
-            let correctPredictions = predictions.argmax(squeezingAxis: 1).elementsEqual(numericLabels)
-            correctGuesses += Int(Tensor<Int32>(correctPredictions).sum())
-            totalGuesses += Int(minibatchSize)
+let (images, numericLabels) = readMNIST(imagesFile: "train-images-idx3-ubyte",
+                                        labelsFile: "train-labels-idx1-ubyte")
+let labels = Tensor<Float>(oneHotAtIndices: numericLabels, depth: 10)
+
+var classifier = Classifier()
+let context = Context(learningPhase: .training)
+let optimizer = SGD<Classifier, Float>(learningRate: 0.2)
+
+// The training loop.
+for epoch in 0..<epochCount {
+    var correctGuessCount = 0
+    var totalGuessCount = 0
+    var totalLoss: Float = 0
+    for i in 0 ..< Int(labels.shape[0]) / batchSize {
+        let x = minibatch(in: images, at: i)
+        let y = minibatch(in: numericLabels, at: i)
+        // Compute the gradient with respect to the model.
+        let 𝛁model = classifier.gradient { classifier -> Tensor<Float> in
+            let ŷ = classifier.applied(to: x, in: context)
+            let correctPredictions = ŷ.argmax(squeezingAxis: 1) .== y
+            correctGuessCount += Int(Tensor<Int32>(correctPredictions).sum().scalarized())
+            totalGuessCount += batchSize
+            let loss = softmaxCrossEntropy(logits: ŷ, labels: y)
+            totalLoss += loss.scalarized()
+            return loss
         }
-        print("""
-              [Epoch \(epoch)] Accuracy: \(correctGuesses)/\(totalGuesses) \
-              (\(Float(correctGuesses) / Float(totalGuesses)))
-              """)
+        // Update the model's differentiable variables along the gradient vector.
+        optimizer.update(&classifier.allDifferentiableVariables, along: 𝛁model)
```

Comment:
Should this be negative?
```suggestion
        optimizer.update(&classifier.allDifferentiableVariables, along: -𝛁model)
```

Reply:
No, because optimizers are responsible for doing minimization given gradients. This is also why `RiemannSGD.update(_:along:)` is doing `.zero - tangent`, speaking of which, the `along:` argument label is not super appropriate.

Reply:
Aha, I see. I agree that `along:` is not quite accurate.